File: legacy/Sallie/backend/services/api-gateway/src/middleware/tracing.ts
import { Request, Response, NextFunction } from 'express';
import { NodeSDK } from '@opentelemetry/sdk-node';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { Resource } from '@opentelemetry/resources';
import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
import { JaegerExporter } from '@opentelemetry/exporter-jaeger';
import { SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { trace, SpanStatusCode } from '@opentelemetry/api';

// Initialize OpenTelemetry
const sdk = new NodeSDK({
  resource: new Resource({
    [SemanticResourceAttributes.SERVICE_NAME]: 'sallie-api-gateway',
    [SemanticResourceAttributes.SERVICE_VERSION]: process.env.npm_package_version || '1.0.0',
    [SemanticResourceAttributes.DEPLOYMENT_ENVIRONMENT]: process.env.NODE_ENV || 'development',
  }),
  instrumentations: [getNodeAutoInstrumentations()],
});

// Initialize Jaeger exporter if configured
if (process.env.JAEGER_ENDPOINT) {
  const exporter = new JaegerExporter({
    endpoint: process.env.JAEGER_ENDPOINT,
  });
  
  sdk.addSpanProcessor(new SimpleSpanProcessor(exporter));
}

// Start the SDK
sdk.start();

// Get tracer
const tracer = trace.getTracer('sallie-api-gateway');

// Middleware to add tracing to requests
export const tracingMiddleware = (req: Request, res: Response, next: NextFunction) => {
  const span = tracer.startSpan(`HTTP ${req.method} ${req.path}`, {
    attributes: {
      'http.method': req.method,
      'http.url': req.url,
      'http.target': req.path,
      'user_agent': req.get('User-Agent') || '',
      'remote_addr': req.ip || req.connection.remoteAddress,
    },
  });

  // Add span to request for access in other middleware/routes
  (req as any).span = span;

  // Track response
  res.on('finish', () => {
    span.setAttributes({
      'http.status_code': res.statusCode,
      'http.status_text': res.statusMessage,
    });

    if (res.statusCode >= 400) {
      span.setStatus({
        code: SpanStatusCode.ERROR,
        message: `HTTP ${res.statusCode}`,
      });
    } else {
      span.setStatus({ code: SpanStatusCode.OK });
    }

    span.end();
  });

  // Handle errors
  res.on('error', (error) => {
    span.recordException(error);
    span.setStatus({
      code: SpanStatusCode.ERROR,
      message: error.message,
    });
    span.end();
  });

  next();
};

// Function to trace service requests
export const traceServiceRequest = async <T>(
  serviceName: string,
  operation: string,
  fn: () => Promise<T>
): Promise<T> => {
  const span = tracer.startSpan(`service.${serviceName}.${operation}`, {
    attributes: {
      'service.name': serviceName,
      'service.operation': operation,
    },
  });

  try {
    const result = await fn();
    
    span.setStatus({ code: SpanStatusCode.OK });
    span.end();
    
    return result;
  } catch (error) {
    span.recordException(error as Error);
    span.setStatus({
      code: SpanStatusCode.ERROR,
      message: (error as Error).message,
    });
    span.end();
    
    throw error;
  }
};

// Function to trace authentication operations
export const traceAuthentication = async <T>(
  operation: string,
  fn: () => Promise<T>
): Promise<T> => {
  const span = tracer.startSpan(`auth.${operation}`, {
    attributes: {
      'auth.operation': operation,
    },
  });

  try {
    const result = await fn();
    
    span.setAttributes({
      'auth.success': true,
    });
    
    span.setStatus({ code: SpanStatusCode.OK });
    span.end();
    
    return result;
  } catch (error) {
    span.setAttributes({
      'auth.success': false,
      'auth.error': (error as Error).message,
    });
    
    span.recordException(error as Error);
    span.setStatus({
      code: SpanStatusCode.ERROR,
      message: (error as Error).message,
    });
    span.end();
    
    throw error;
  }
};

// Function to add custom events to spans
export const addSpanEvent = (req: Request, eventName: string, attributes?: Record<string, any>) => {
  const span = (req as any).span;
  if (span) {
    span.addEvent(eventName, attributes);
  }
};

// Function to set span attributes
export const setSpanAttribute = (req: Request, key: string, value: any) => {
  const span = (req as any).span;
  if (span) {
    span.setAttribute(key, value);
  }
};

// Export tracer for use in other modules
export { tracer };

// Graceful shutdown
process.on('SIGTERM', () => {
  sdk
    .shutdown()
    .then(() => console.log('OpenTelemetry shut down successfully'))
    .catch((error) => console.error('Error shutting down OpenTelemetry', error))
    .finally(() => process.exit(0));
});

process.on('SIGINT', () => {
  sdk
    .shutdown()
    .then(() => console.log('OpenTelemetry shut down successfully'))
    .catch((error) => console.error('Error shutting down OpenTelemetry', error))
    .finally(() => process.exit(0));
});

File: legacy/Sallie/backend/services/python-ai-service/src/middleware/tracing.py
from fastapi import Request
from starlette.middleware.base import BaseHTTPMiddleware
from opentelemetry import trace
from opentelemetry.exporter.jaeger.thrift import JaegerExporter
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.httpx import HTTPXClientInstrumentor
from opentelemetry.sdk.resources import SERVICE_NAME, RESOURCE
import logging
from ..core.config import settings

logger = logging.getLogger(__name__)

class TracingMiddleware(BaseHTTPMiddleware):
    """Middleware for OpenTelemetry tracing"""
    
    def __init__(self, app):
        super().__init__(app)
        self._setup_tracing()
    
    def _setup_tracing(self):
        """Setup OpenTelemetry tracing"""
        try:
            if settings.JAEGER_ENDPOINT:
                # Configure Jaeger exporter
                jaeger_exporter = JaegerExporter(
                    agent_host_name="localhost",
                    agent_port=6831,
                    endpoint=settings.JAEGER_ENDPOINT,
                )
                
                # Configure tracer provider
                trace.set_tracer_provider(
                    TracerProvider(
                        resource=RESOURCE.create(
                            {
                                SERVICE_NAME: settings.APP_NAME,
                                SERVICE_VERSION: settings.APP_VERSION,
                            }
                        )
                    )
                )
                
                # Add span processor
                tracer_provider = trace.get_tracer_provider()
                span_processor = BatchSpanProcessor(jaeger_exporter)
                tracer_provider.add_span_processor(span_processor)
                
                # Instrument FastAPI and HTTPX
                FastAPIInstrumentor.instrument_app(self.app)
                HTTPXClientInstrumentor.instrument()
                
                logger.info("OpenTelemetry tracing initialized")
            else:
                logger.info("Jaeger endpoint not configured, tracing disabled")
        except Exception as e:
            logger.error(f"Failed to setup tracing: {e}")
    
    async def dispatch(self, request: Request, call_next):
        """Add tracing to requests"""
        tracer = trace.get_tracer(__name__)
        
        with tracer.start_as_current_span(
            f"{request.method} {request.url.path}",
            kind=trace.SpanKind.SERVER,
            attributes={
                "http.method": request.method,
                "http.url": str(request.url),
                "http.target": request.url.path,
                "http.host": request.url.hostname,
                "http.scheme": request.url.scheme,
                "user.agent": request.headers.get("user-agent", ""),
            },
        ) as span:
            try:
                response = await call_next(request)
                
                span.set_attributes({
                    "http.status_code": response.status_code,
                })
                
                if response.status_code >= 400:
                    span.set_status(trace.Status.ERROR)
                else:
                    span.set_status(trace.Status.OK)
                
                return response
                
            except Exception as e:
                span.set_status(trace.Status.ERROR)
                span.record_exception(e)
                raise

# Function to get current span
def get_current_span():
    """Get the current OpenTelemetry span"""
    return trace.get_current_span()

# Function to add attributes to current span
def add_span_attribute(key: str, value: str):
    """Add attribute to current span"""
    span = get_current_span()
    if span:
        span.set_attribute(key, value)

# Function to add event to current span
def add_span_event(name: str, attributes: dict = None):
    """Add event to current span"""
    span = get_current_span()
    if span:
        span.add_event(name, attributes or {})
